### ISSUE TITLE
Accept import of plugin with ver. same as already installed

### ISSUE THREAD

**Issue by twMat** [TiddlyWiki/TiddlyWiki5#2441]
There are occasions when wanting to ensure that a plugin is not messed up. Particularly in developing plugins it is desirable to be able to easily test import a plugin "as if a user". However, import mechanism blocks importing of same plugin with same version, thus forcing manual deletion of the already installed one first.

''Proposal'': Permit import (overwrite) of plugins with version same as already installed.

**Comment by sukima** [comment]
Why not use a build number / SHA as part of the version. [Semantic Versioning][1] would offer _in development_ extentions:

> Additional labels for pre-release and build metadata are available as extensions to the MAJOR.MINOR.PATCH format.

So you could make your versions look like: `1.2.3.1234` (build number), `1.2.3-alpha3` alpha/beta tags, or `1.2.3-bada55` Git SHA for development.

This way the version is always unique but doesn't violate [semver][1]. Then when your ready for a release bump the version as you would normally (`1.2.4` / `1.3.0` / `2.0.0`).

[1]: http://semver.org/

**Comment by pmario** [comment]
I think this isn't a technical problem, but just about the "users feelings". If I have a `--force install` button, I can be sure that everything is overwritten and there is no need to check again. Because the systems "don't overwrite user content" is not active.

The problem we have is, that we would need to delete every user modified tiddler, that is part of the plugin too!

**Comment by twMat** [comment]
Hm, both your answers make me realize there's a presumption on my side;

IMO one should not be forced outside of TW to build TW stuff, at least not unless it's really complex stuff. That's pretty much what TW is about IMO; build stuff with *with/from/of* TW itself. One categorical example of stuff to build is "native TW plugins". By "native" I mean plugins that consist of native TW components (plus perhaps html/css). In other words, no external code environment and not even node-js TW should be necessary for this. But it strikes me that this might not be self evident? (Would love to hear @Jermolene 's take on this!)
...

So, @sukima - "build number / SHA" - I guess I'll have to bring this suggestion up with Tinka maker @Drakor because Tinka is the only way I know for how to make plugins.

@pmario - "force install" - any suggestion for how to do this in standard TW?

>The problem we have is, that we would need to delete every user modified tiddler, that is part of the plugin too!

I don't understand this. Is a plugin not actually just a single big tiddler where modified shadow tiddlers are read instead of their respective sections in that big tiddler? I cannot locate any connection between a modified shadow and the *version* of the original plugin, so why would mods need to be deleted?

**Comment by pmario** [comment]
>> The problem we have is, that we would need to delete every user modified tiddler, that is part of the plugin too!

>I don't understand this.

There is **only one** valid reason, that I can think of, which would make your request needed. ... 

You want to force a clean install of a plugin, because someone may have changed it, but you don't know. 

>--- from OP ----
There are occasions when wanting to ensure that a plugin is not messed up. Particularly in developing plugins it is desirable to be able to easily test import a plugin "as if a user". 

So there may be user modified shadow tiddlers, that need to be overwritten by the `forced` plugin. 

Every other reason just means, there is a bug in the SW or the workflow.

**Comment by twMat** [comment]
I appreciate your input but in **my** use case I wish to renew the plugin but with the same version number. This is *during dev*. The "messing up" refers to myself. I want to be able to create a plugin, test install it in TW n (as if I were a user), perhaps tweak the plugin and then test install it again in TW n, without having to first manually remove the plugin from TW n.

In this context, overwritten shadows actually allow me to easily perform tests (in TW n) so they would better not be force deleted - it is very easy to delete them anyway to switch to the refreshed shadow.

**Comment by Drakor** [comment]
_Quick "Off-Topic" Info for Mat:_

Tinka actually bumps up the PATCH part of the version number whenever you make a change with the main goal of it being able to be installed over the current version (and it also gets rid of possible shadow tiddlers that could overwrite plugin tiddlers).

While TW actually does this automatically, what it allows you to do is tweak a plugin, package the  changes and have those immediately loaded (that's why it also forces you to refresh the wiki). Furthermore, if you made a mistake during development, you can quickly revert by "restoring" the previous "development version" of the plugin via the "Archive" tab since Tinka backups your plugin every time you package changes into it. Note however that this will not delete unpackaged tiddlers, i.e. modified shadow tiddlers that overwrite plugin tiddlers. 

If you want to get rid of unpackaged changes you will have to delete the modified tiddlers either manually or do a "Quick-Package" of the plugin via the help tab (which will incorporate the bad changes into a new version) and then restore the previous version of the plugin. This process will work, however it is also both confusing and will lead to a bad backup of the changes that you wanted  to get rid of. Therefore I think your use case is incorporated in the current functionality as long as you don't want the modified shadow tiddlers to be overwritten too, if that's the case, let me know and I might add a "Clean Plugin Environment" or "Clean State" button.

**Comment by twMat** [comment]
Doesn't seem to be a PM tool here so answering publically;
@Drakor - thanks for explanation! Your points are of the kind that I have to try things in parallel to reading them. Will do so on next packing occasion. As I've said before; I think your creation to a great deal solves a truly important infrastructurial matter in the TW community - but I also feel it needs some further polishing. For next use, I hope to document the exact stumbles i hit including what we''ve touched here.

**Comment by pmario** [comment]
- The [innerwiki plugin](https://tiddlywiki.com/#Innerwiki%20Plugin) can handle this
- Inner wiki usage: [Javascript Widget Tutorial](https://tiddlywiki.com/dev/#Javascript%20Widget%20Tutorial)

**Comment by twMat** [comment]
@pmario 

> The [innerwiki plugin](https://tiddlywiki.com/#Innerwiki%20Plugin) can handle this

Come on - one shouldn't need a totally separate plugin for this! This issue still bothers me but it is even reasonable to also expect regular end users, who are not plugin developers, to want to be able to install/overwrite the same version of a plugin perhaps to test if a current plugin instance is somehow faulty. There may not even *be* a higher version so then the user is forced to *uninstall* the plugin first etc. 

Please reopen. I apparently don't have the rights to do so. Or are there any strong arguments *against* this OP?

**Comment by pmario** [comment]
There is a strong argument to not import a plugin with the exact same version number. Because it is already there. If even 1 typo in the documentation is changed, there has to be a new version number.

- There will be a user, that will point out, that eg: v1.1.1 has a typo in the docs, which you had fixed a month ago with the same version number
- Same version means **exactly the same content** otherwise version numbers make no sense
- If you ever forget to update the version number and publish it, your users will have problems
- You will have maintenance problems with errors which are almost impossible to find
  - Eg: If you mess with a plugin, there only has to be a telephone call or any other distraction longer than several minutes and you can not be sure anymore what exactly has been changed.

**Comment by twMat** [comment]
@pmario 

You argument puts me "on the fence" for this.

But so what should be done? To manually have to delete a plugin before importing a tweaked one (rather than just importing directly) is very annoying when it has to be done repeatedly. How wold you suggest to simplify this in a smooth way?

Also, here are two tiddlers, are they identical?

```
title:foo
version: 1
modified: .....1
```

```
title:foo
version: 1
modified: .....2
```

**Comment by pmario** [comment]
While I did create the Lucide icon plugin creator, I did add some new functionality to the bundler plugin. It now can create plugins with an action-bundleplugin widget. Per default it does an auto-increment. 

For the Lucide workflow, I did create a "preview" which shows a log-file, what should be done but it does not modify the plugin

The preview also allows to manually define a completely new version. If nothing is changed, the "patch" is auto-incremented.

I think, if it is implemented in the workflow and everything "falls into place" it should be convenient instead of annoying.

The problem at the moment is, that it is not documented yet, since I was not sure, if will keep the workflow.

**Comment by saqimtiaz** [comment]
> But so what should be done? To manually have to delete a plugin before importing a tweaked one (rather than just importing directly) is very annoying when it has to be done repeatedly.

I think it is a extremely minute fraction of users that encounter this issue, and that we should not necessarily add complexity to the core to workaround this, considering that for the majority of users the current behaviour is a sensible default.

 Perhaps a custom upgrader module could be an option, or even a a custom dropzone (entirely wikitext based), for those that need this workflow frequently.

**Comment by twMat** [comment]
> I think it is a extremely minute fraction of users that encounter this issue, and that we should not necessarily add complexity to the core

Right. I was hoping for some very simple solution . It is not a *big* deal to me, merely a recurring annoyance.

@pmario , if I understand your last post you're bringing up plugin creator tools that feature auto-increment. I use @Drakor 's Tinka plugin packer which also features this but the issue is in the importing mechanism as described in the OP.

Anyway, I won't push the issue further and keep on with the annoyance.

**Comment by pmario** [comment]
>@pmario , if I understand your last post you're bringing up plugin creator tools that feature auto-increment. 

Yes.

>I use @Drakor 's Tinka plugin packer which also features this but the issue is in the importing mechanism as described in the OP.

That's right, but if you test new plugin features and the "compiler" auto-increments, than the import should cause no problems.
